### ISSUE TITLE
Acces to database from outside in k8s

### ISSUE THREAD

**Issue by CrazyMegaDuck** [zalando/postgres-operator#1216]
Hello!
I have installed postgres operator. How can I get access to postgres database from outside? I tried to change cluster service type from ClusterIP to NodePort, but it is overwritten automatically.

**Comment by xzizka** [comment]
For testing purposes, you can use port-forward e.g.
`kubectl port-forward <service-name> 5432:5432` 
Have you tried to create a new service instead of editing the one created by postgres-operator?

**Comment by FxKu** [comment]
With `enable_master_load_balancer: true` you can change the service to type `LoadBalancer`. Then it's up to your setup to provide an accessible endpoint.

**Comment by CrazyMegaDuck** [comment]
> For testing purposes, you can use port-forward e.g.
> `kubectl port-forward <service-name> 5432:5432`
> Have you tried to create a new service instead of editing the one created by postgres-operator?
Yeah, this is a quick way to test. But I have to give access to all users on the local network.

**Comment by CrazyMegaDuck** [comment]
> With `enable_master_load_balancer: true` you can change the service to type `LoadBalancer`. Then it's up to your setup to provide an accessible endpoint.

Thank you very much! It works!

**Comment by timmy59100** [comment]
@FxKu This works for me but on the master service the selector labels are not created. On the replica service the selector labels are there. So I have to add them myself to access the pod.

Or am I'm doing something wrong here?

  selector:
    application: spilo
    cluster-name: CLUSTERNAME
    spilo-role: master

**Comment by yanchenko-igor** [comment]
> @FxKu This works for me but on the master service the selector labels are not created. On the replica service the selector labels are there. So I have to add them myself to access the pod.
> 
> Or am I'm doing something wrong here?
> 
> selector:
> application: spilo
> cluster-name: CLUSTERNAME
> spilo-role: master

Which version of postgres-operator/spilo are you using? I had the same issue when I upgraded to kubernetes 1.19, but it was solved in the latest versions of postgres-operator/spilo.

**Comment by timmy59100** [comment]
@yanchenko-igor 
Thanks. I'm on Kubernetes 1.19 too. I just updated from spilo-12:1.6-p3 to spilo-12:1.6-p5 but the problem ist persistent.

**Comment by yanchenko-igor** [comment]
That one is too old, I am using `registry.opensource.zalan.do/acid/spilo-cdp-13:2.0-p146` but `registry.opensource.zalan.do/acid/spilo-cdp-12:2.0-p144` should be fine too.

**Comment by timmy59100** [comment]
Ah, I didn't find these. Where are the versions listed?
Updated to registry.opensource.zalan.do/acid/spilo-cdp-13:2.0-p146. But it didn't fix my issue.

**Comment by yanchenko-igor** [comment]
Service should be changed, patroni operates with endpoints directly, it should just work.

**Comment by timmy59100** [comment]
Yes, now I have access to the service.
Thanks!

**Comment by timmy59100** [comment]
@yanchenko-igor @FxKu 
On the main page the image is listed as: registry.opensource.zalan.do/acid/spilo-13:2.0-p2
What is the difference to registry.opensource.zalan.do/acid/spilo-cdp-13:2.0-p146, which should I choose?

Maybe always the latest, which is now registry.opensource.zalan.do/acid/spilo-cdp-13:2.0-p149?

**Comment by FxKu** [comment]
`spilo-13:2.0-p2` is a released version.
`spilo-cdp-13:2.0-p149` is the latest dev build. The last release is equivalent to the `p147` so there's not a big deviation atm. And I think with some extensions currently updated (Timescale, PostGIS) we will have another release pretty soon.

**Comment by tylerrhinehart** [comment]
> Service should not be changed, patroni operates with endpoints directly, it should just work.

@yanchenko-igor Can you elaborate? What steps or commands do I need to run to expose the service/endpoints outside the cluster?

**Comment by yanchenko-igor** [comment]
> > Service should not be changed, patroni operates with endpoints directly, it should just work.
> 
> @yanchenko-igor Can you elaborate? What steps or commands do I need to run to expose the service/endpoints outside the cluster?

No need to do anything, patroni will do the job.

**Comment by tylerrhinehart** [comment]
> No need to do anything, patroni will do the job.

@yanchenko-igor Then how do I access patroni from outside the cluster? The only thing I have found to work is manually creating a new service.

**Comment by ru5t** [comment]
@tylerrhinehart
If you create cluster using https://postgres-operator.readthedocs.io/en/latest/quickstart/ , then you use `manifests/minimal-postgres-manifest.yaml`, which doesn't have enableMasterLoadBalancer enabled, so no LoadBalancer service created.

If you use `operator ui`, then you can enable this setting by using checkbox `Enable master ELB`. But in this case you get unaccessible service. Issue here is that `new cluster configuration` wizard doesn't let you set `allowedSourceRanges`, even though it shown in config preview on the left side of page. So you'll have LoadBalancer service, but not much use of it. 

I'm new to Patroni/ Postgres-operator, so faced this issue during setup of test kind cluster. Found this thread looking for help for same problem. It seems, it wasn't the case before - https://github.com/zalando/postgres-operator/issues/54 . It looks like security measure.

So, solution is to use `manifests/minimal-postgres-manifest.yaml` with added spec.enableMasterLoadBalancer: true and with `allowedSourceRanges` set depending of your environment.

Hope `operator-ui` would be updated to provide way for  `allowedSourceRanges` mods.

**Comment by Vivida1** [comment]
So if I cannot use the type LoadBalancer option (because my Cluster does not support it) and since I cannot create services which use a label selector (since Patroni modifies Endpoints directly), there is basically no chance for me to expose the cluster?

Wouldn't it be possible to add the option to create the existing service as type NodePort instead of ClusterIP?

**Comment by marcellodesales** [comment]
Also, shouldn't `enableMasterLoadBalancer: true` be a result of `numberOfInstances > 1`???

**Comment by UPiotr** [comment]
> @tylerrhinehart If you create cluster using https://postgres-operator.readthedocs.io/en/latest/quickstart/ , then you use `manifests/minimal-postgres-manifest.yaml`, which doesn't have enableMasterLoadBalancer enabled, so no LoadBalancer service created.
> 
> If you use `operator ui`, then you can enable this setting by using checkbox `Enable master ELB`. But in this case you get unaccessible service. Issue here is that `new cluster configuration` wizard doesn't let you set `allowedSourceRanges`, even though it shown in config preview on the left side of page. So you'll have LoadBalancer service, but not much use of it.
> 
> I'm new to Patroni/ Postgres-operator, so faced this issue during setup of test kind cluster. Found this thread looking for help for same problem. It seems, it wasn't the case before - #54 . It looks like security measure.
> 
> So, solution is to use `manifests/minimal-postgres-manifest.yaml` with added spec.enableMasterLoadBalancer: true and with `allowedSourceRanges` set depending of your environment.
> 
> Hope `operator-ui` would be updated to provide way for `allowedSourceRanges` mods.

Thank you!!!!!

I was missing allowedSourceRanges and was wondering why my LB isn't working. Completely forgot about pg_hba.conf
